include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+                region.imageExtent = params.extent;
+                auto buffer = core::make_smart_refctd_ptr<ICPUBuffer>(getTexelOrBlockBytesize(_format) * params.extent.width*params.extent.height*params.arrayLayers);
+                image->setBufferAndRegions(std::move(buffer), regions);
+            }
+        }
+
+    private:
+        core::smart_refctd_ptr<ICPUImageView> createView_internal(ICPUImageView::SCreationParams&& _params) const override
+        {
+            return ICPUImageView::create(std::move(_params));
+        }
+    };
+
+    ICPUVirtualTexture(
+        const base_t::IVTResidentStorage::SCreationParams* _residentStorageParams,
+        uint32_t _residentStorageCount,
+        uint32_t _pgTabSzxy_log2 = 8u,
+        uint32_t _pgTabLayers = 256u,
+        uint32_t _pgSzxy_log2 = 7u,
+        uint32_t _tilePadding = 9u,
+        uint32_t _maxAllocatableTexSz_log2 = 14u
+    ) : IVirtualTexture(
+        _pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _tilePadding
+    ) {
+        m_pageTable = createPageTable(_pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _maxAllocatableTexSz_log2);
+        initResidentStorage(_residentStorageParams, _residentStorageCount);
+    }
+
+    STextureData pack(const ICPUImage* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        uint32_t smplrIndex = 0u;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return STextureData::invalid();
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+
+            SamplerArray* views = nullptr;
+            if (isFloatingPointFormat(format)||isNormalizedFormat(format)||isScaledFormat(format))
+                views = &m_fsamplers;
+            else if (isSignedFormat(format))
+                views = &m_isamplers;
+            else
+                views = &m_usamplers;
+            auto view_it = std::find_if(views->views->begin(), views->views->end(), [format](const core::smart_refctd_ptr<ICPUImageView>& _view) {return _view->getCreationParameters().format==format;});
+            if (view_it==views->views->end()) //no physical page texture view/sampler for requested format
+                return STextureData::invalid();
+            smplrIndex = std::distance(views->views->begin(), view_it);
+        }
+        auto assignedLayers = storage->getPageTableLayers();
+
+        page_tab_offset_t pgtOffset = page_tab_offset_invalid();
+        for (auto it = assignedLayers.first; it != assignedLayers.second; ++it)
+        {
+            pgtOffset = alloc(_img, _subres, *it);
+            if ((pgtOffset==page_tab_offset_invalid()).all())
+                continue;
```

Comment:
I think you meant `break`

Reply:
and a `!` in the `if`

Reply:
no, im iterating through all layers assigned to this storage. If i cant allocate on `it` then trying next one

Reply:
and if you don't fail you try the next one anyway because its a `for (auto it = assignedLayers.first; it != assignedLayers.second; ++it)` loop

Also the logic is off, because layers are assigned to formats (views) not format classes (storage)

This is why I essntially propose to get rid of both assigned layers in storage class, and the free layer stack, just use a `unordered_multimap<E_FORMAT,uint32_t>` plus the layer to format sampler view LUT